ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Util.hs

```diff
@@ -374,3 +388,51 @@ fib n = round $ phi ** fromIntegral n / sq5
 eitherToMaybe :: Either a b -> Maybe b
 eitherToMaybe (Left _)  = Nothing
 eitherToMaybe (Right x) = Just x
+
+{-------------------------------------------------------------------------------
+  Fuse code, i.e. regions of code that will throw an exception if accessed
+  concurrently.
+-------------------------------------------------------------------------------}
+
+-- | A WithIntactFuse expression is one of which a single instance can be
+-- executed at any given moment in the lifetime of the program.
+newtype WithIntactFuse m a = WithIntactFuse (m a)
+  deriving newtype (Functor, Applicative, Monad, MonadThrow, MonadCatch)
+
+instance MonadTrans WithIntactFuse where
+  lift = WithIntactFuse
+
+-- | Declare that an action is not safe to be run concurrently, i.e. two
+-- concurrent accesses should throw an exception.
+blowFuse :: m a -> WithIntactFuse m a
+blowFuse = WithIntactFuse
+
+-- | Lock required to run a 'WithIntactFuse'
+data Fuse m = Fuse String (MVar m ())
+
+newFuse :: MonadMVar m => String -> m (Fuse m)
+newFuse name = Fuse name <$> newMVar ()
+
+-- | Run the 'WithIntactFuse' action throwing an exception if it is already
+-- running in a separate thread.
+--
+-- Note that if the 'WithIntactFuse' action throws an exception, the fuse will
+-- not be resetted, and any subsequent attempts to use it will fail.
+blowAndResetFuse ::
+     (MonadThrow m, MonadMVar m)
+  => Fuse m
+  -> WithIntactFuse m a
+  -> m a
+blowAndResetFuse (Fuse name m) (WithIntactFuse io) = do
+    onlyOne =<< tryTakeMVar m
+    a <- io
+    onlyOne . bool Nothing (Just ()) =<< tryPutMVar m ()
```

Comment:
Minor: I think `bool` is poor for legibility: it's not obvious which argument is which.

Reply:
Also, I think we should throw a distinct exception when the write fails. Because that means some other code somehow entered when it shouldn't have, which would be debugged differently?

Reply:
Also, in my opinion, it'd be easier to grok the implementation if you inlined `onlyOne` and simplified down to something like:

```
tryTakeMVar m >>= \case
  Nothing -> throwIO $ ...
  Just () -> pure ()

...
  
tryPutMVar m () >>= \b -> unless b $ throwIO $ ...
```